debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/converters/sink/types/bson/DecimalType.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.connector.mongodb.converters.sink.types.bson;
+
+import java.math.BigDecimal;
+
+import org.apache.kafka.connect.data.Decimal;
+import org.apache.kafka.connect.errors.DataException;
+import org.bson.BsonDecimal128;
+import org.bson.BsonDouble;
+import org.bson.BsonValue;
+import org.bson.types.Decimal128;
+
+public class DecimalType extends AbstractBsonType {
+
+    public enum Format {
+        DECIMAL128, // for MongoDB v3.4+
+        LEGACYDOUBLE // results in double approximation
+    }
+
+    private final Format format;
+
+    public DecimalType() {
+        super(Decimal.schema(0));
+        this.format = Format.DECIMAL128;
+    }
+
+    public DecimalType(final Format format) {
+        super(Decimal.schema(0));
+        this.format = format;
+    }
+
+    @Override
+    public BsonValue toBson(final Object data) {
+        if (data instanceof BigDecimal) {
+            if (format.equals(Format.DECIMAL128)) {
+                return new BsonDecimal128(new Decimal128((BigDecimal) data));
+            }
+            if (format.equals(Format.LEGACYDOUBLE)) {
+                return new BsonDouble(((BigDecimal) data).doubleValue());
+            }
```

Comment:
Another if/elseif or perhaps switch use case.

Reply:
Same response as in other cases :)... +1 for switch

Reply:
filed https://issues.redhat.com/browse/DBZ-8006 for the Java 17 enhancements/benefits

Reply:
Leave it as is now and add it to the refactorig Jira.